File: app/src/main/java/com/pietrantuono/podcasts/subscribedpodcasts/presenter/SubscribedPodcastPresenter.java
package com.pietrantuono.podcasts.subscribedpodcasts.presenter;

import com.pietrantuono.podcasts.GenericPresenter;
import com.pietrantuono.podcasts.addpodcast.model.pojos.SinglePodcast;
import com.pietrantuono.podcasts.subscribedpodcasts.model.SubscribedPodcastModel;
import com.pietrantuono.podcasts.subscribedpodcasts.view.SubscribedPodcastView;

import java.util.List;

import rx.Observer;

public class SubscribedPodcastPresenter implements GenericPresenter {
    public static final String TAG = SubscribedPodcastPresenter.class.getSimpleName();
    private SubscribedPodcastView view;
    private final SubscribedPodcastModel model;

    public SubscribedPodcastPresenter(SubscribedPodcastModel model) {
        this.model = model;
    }

    public void bindView(SubscribedPodcastView view) {
        this.view = view;
    }

    public void onResume() {
        model.subscribeToSubscribedPodcasts(new Observer<List<SinglePodcast>>() {
            @Override
            public void onCompleted() {

            }

            @Override
            public void onError(Throwable throwable) {
                view.onError(throwable);
            }

            @Override
            public void onNext(List<SinglePodcast> singlePodcasts) {
                if (singlePodcasts != null || !singlePodcasts.isEmpty()) {
                    view.setPodcasts(singlePodcasts);
                }
            }
        });
    }

    @Override
    public void onDestroy() {

    }

    public void onPause() {
        model.unsubscribe();
    }
}

File: app/src/main/java/com/pietrantuono/podcasts/subscribedpodcasts/presenter/SubscribedPodcastPresenter.kt
package com.pietrantuono.podcasts.subscribedpodcasts.presenter

import android.arch.lifecycle.ViewModel
import android.widget.ImageView
import android.widget.LinearLayout
import com.pietrantuono.podcasts.GenericPresenter
import com.pietrantuono.podcasts.addpodcast.customviews.PodcastsAdapter
import com.pietrantuono.podcasts.addpodcast.view.ApiLevelChecker
import com.pietrantuono.podcasts.singlepodcast.presenter.SimpleObserver
import com.pietrantuono.podcasts.subscribedpodcasts.model.SubscribedPodcastModel
import com.pietrantuono.podcasts.subscribedpodcasts.view.SubscribedPodcastView
import models.pojos.Podcast

class SubscribedPodcastPresenter(
        private val model: SubscribedPodcastModel,
        private val apiLevelChecker: ApiLevelChecker) :
        GenericPresenter, PodcastsAdapter.OnItemClickedClickedListener, ViewModel() {

    override fun onDestroy() {}

    override fun onPause() {
        model.unsubscribe()
    }

    private var view: SubscribedPodcastView? = null

    fun bindView(view: SubscribedPodcastView?) {
        this.view = view
    }

    override fun onResume() {
        model.subscribeToSubscribedPodcasts(object : SimpleObserver<List<Podcast>>() {
            override fun onError(throwable: Throwable) {
                view?.onError(throwable)
            }

            override fun onNext(podcasts: List<Podcast>?) {
                if (podcasts != null || !podcasts!!.isEmpty()) {
                    view?.setPodcasts(podcasts)
                }
            }
        })
    }

    override fun onItemClicked(podcast: Podcast?, imageView: ImageView?, position: Int,
                               titleContainer: LinearLayout?) {
        if (apiLevelChecker.isLollipopOrHigher && !isViewPartiallyHidden(position)) {
            view?.startDetailActivityWithTransition(podcast, imageView, titleContainer)
        } else {
            view?.startDetailActivityWithoutTransition(podcast)
        }
    }

    companion object {
        val TAG = SubscribedPodcastPresenter::class.java.simpleName
    }

    private fun isViewPartiallyHidden(position: Int): Boolean {
        if (view != null) {
            return view!!.isPartiallyHidden(position)
        } else {
            return false
        }
    }
}
